cram/cram_encode.c

```diff
@@ -3852,6 +3904,28 @@ int cram_put_bam_seq(cram_fd *fd, bam_seq_t *b) {
             c->multi_seq = 1;
             c->pos_sorted = 0;
 
+            // Cram_next_container may end up flushing an existing one and
+            // triggering fd->embed_ref=2 if no reference is found.
+            // Embedded refs are incompatible with multi-seq, so we bail
+            // out and switch to no_ref in this scenario.  We do this
+            // within the container only, as multi_seq may be temporary
+            // and we switch back away from it again.
+            pthread_mutex_lock(&fd->ref_lock);
+            if (embed_ref + c->embed_ref + fd->embed_ref != 0)
+            if (fd->embed_ref > 0 && c->curr_rec == 0 && c->curr_slice == 0) {
```

Comment:
The indentation of these `if` statements looks dubious.  Are they both supposed to be here?

Reply:
A bit odd I'll agree and definitely not intended as-is.

When `c->curr_rec == 0 && c->curr_slice == 0` at the start of a slice, `embed_ref`, `c->embed_ref` and `fd->embed_ref` are all the same thing.  So the former statement is redundant I think.  Maybe it was an alternative statement from an earlier incarnation.  I sometimes had two variants and comment one or the other out to compare how they work, so it's probably a failure to tidy up properly.

It does make me wonder whether this works correctly when swapping from one slice to another in a multi-slice containers, but you've got to really go through hoops to explore this scenario as it can never happen automatically.  It's complex because the reference ID is stored in the container header, but the local ref MD5sum and/or embedded ref block number are stored in the slice header.  The whole unsorted detection though is done at the container level as it's that which stores ref start/end.

The whole multi-slice thing was probably a mistake in CRAM tbh, and it dates back to an earlier design in CRAM 1.x.  Even Vadim had stopped using it by the time I came along, so now it's just baggage!  (Unless we redesign the codecs to put more frequency meta-data into the container compression header.)